ports/esp32/esp32_can.c

```diff
@@ -0,0 +1,759 @@
+// INTERNAL FUNCTION Return status information
+STATIC twai_status_info_t _esp32_hw_can_get_status() {
+    twai_status_info_t status;
+    check_esp_err(twai_get_status_info(&status));
+    return status;
+}
+
+// INTERNAL FUNCTION Reset can filter to defaults
+void esp32_reset_can_filter(const esp32_can_obj_t *self) {
+    twai_filter_config_t f_config = TWAI_FILTER_CONFIG_ACCEPT_ALL();
+    self->config->filter.single_filter = f_config.single_filter;
+    self->config->filter.acceptance_code = f_config.acceptance_code;
+    self->config->filter.acceptance_mask = f_config.acceptance_mask;
+}
+
+STATIC void esp32_hw_can_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    esp32_can_obj_t *self = MP_OBJ_TO_PTR(self_in);
+
+    if (self->config->initialized) {
+        qstr mode = MP_QSTR_LISTEN;
+        switch (self->config->general.mode) {
+            case CAN_MODE_LOOPBACK:
+                mode = MP_QSTR_LOOPBACK;
+                break;
+            case CAN_MODE_SILENT:
+                mode = MP_QSTR_SILENT;
+                break;
+            case CAN_MODE_NORMAL:
+                mode = MP_QSTR_NORMAL;
+                break;
+        }
+        mp_printf(print, "CAN(tx=%u, rx=%u, baudrate=%ukb, mode=%q, loopback=%u)",
+            self->config->general.tx_io,
+            self->config->general.rx_io,
+            self->config->baudrate,
+            mode,
+            self->loopback
+            );
+    } else {
+        mp_printf(print, "Device is not initialized");
+    }
+}
+
+// INTERNAL FUNCTION FreeRTOS IRQ task
+STATIC void esp32_hw_can_irq_task(void *self_in) {
+    esp32_can_obj_t *self = (esp32_can_obj_t *)self_in;
+    uint32_t alerts;
+
+    twai_reconfigure_alerts(
+        TWAI_ALERT_RX_DATA | TWAI_ALERT_RX_QUEUE_FULL | TWAI_ALERT_BUS_OFF | TWAI_ALERT_ERR_PASS |
+        TWAI_ALERT_ABOVE_ERR_WARN | TWAI_ALERT_TX_FAILED | TWAI_ALERT_TX_SUCCESS | TWAI_ALERT_BUS_RECOVERED,
+        NULL
+        );
+
+    while (1) {
+        check_esp_err(twai_read_alerts(&alerts, portMAX_DELAY));
+
+        if (alerts & TWAI_ALERT_BUS_OFF) {
+            ++self->num_bus_off;
+        }
+        if (alerts & TWAI_ALERT_ERR_PASS) {
+            ++self->num_error_passive;
+        }
+        if (alerts & TWAI_ALERT_ABOVE_ERR_WARN) {
+            ++self->num_error_warning;
+        }
+
+        if (alerts & (TWAI_ALERT_TX_FAILED | TWAI_ALERT_TX_SUCCESS)) {
+            self->last_tx_success = (alerts & TWAI_ALERT_TX_SUCCESS) > 0;
+        }
+
+        if (alerts & (TWAI_ALERT_BUS_RECOVERED)) {
+            self->bus_recovery_success = true;
+        }
+
+        if (self->rxcallback != mp_const_none) {
+            if (alerts & TWAI_ALERT_RX_DATA) {
+                uint32_t msgs_to_rx;
+                msgs_to_rx = _esp32_hw_can_get_status().msgs_to_rx;
+
+                if (msgs_to_rx == 1) {
+                    // first message in queue
+                    mp_sched_schedule(self->rxcallback, mp_obj_new_int(0));
```

Comment:
The identifier that is passed in the callback should be a constant that is exposed to the python side of things. 

You are using `mp_obj_new_int(0)` in this line and then in the lines below 1 and 2 are being done the same way. It is not going to be know what those values mean.

I think that it would be best to pass the actual message that is received and open up that receive buffer asap. You can collect the message information and pass it as a dictionary. This way you wouldn't have to bother with the 3 different integer values that are being passed to the callback. I would actually empty all of the receive buffers and pass an iterable of dictionaries. if a callback is received for TWAI_ALERT_RX_QUEUE_FULL then empty the buffer.

This would obviously only apply if the user registered a callback. The user should be able to single out what they want callbacks for.  It look like having a callback is going to be a requirement in your code because there is no checking of rxcallback

Reply:
> The identifier that is passed in the callback should be a constant that is exposed to the python side of things.

Do you mean I should use `MP_OBJ_NEW_SMALL_INT` instead if `mp_obj_new_int`? That is my fault, I'll fix it. It could also be good idea to put values (0, 1, 2) to some constant or definitions

> You can collect the message information and pass it as a dictionary.

I would like to make it in this way, but this API interface is made as close as possible to `pyb.CAN` and according to [pyb.CAN rxcallback documentation](https://docs.micropython.org/en/latest/library/pyb.CAN.html#pyb.CAN.rxcallback) `reason` argument should be one of `0`, `1`, or `2` and python-side code should look like 

```python
def cb0(bus, reason):
  if reason == 0:
      print('pending')
  if reason == 1:
      print('full')
  if reason == 2:
      print('overflow')
```

So there is no way (until massive refactoring and api-breaking changes) to make it.

Reply:
I don't believe the STM32 way of handling the callback is the best way to go about doing it. It is dependent on what appears to be a scheduler lock which I would think is going to unlock when it has time to. I would think that would be holding up the interrupt  from being released. It also looks like the callback is made from inside the interrupt handler.  until it makes the callback. It is also letting the buffer sit there and not get emptied. How large is the receive buffer is it large enough to wait to be emptied?.

If the interrupt pulled the message and passed to to the scheduler to make the call to the callback when here is time to tht would empty the buffer freeing up space and release the interrupt.


But at any rate..

This is the callback handler for the STM32

```c
STATIC void can_rx_irq_handler(uint can_id, uint fifo_id) {
    mp_obj_t callback;
    pyb_can_obj_t *self;
    mp_obj_t irq_reason = MP_OBJ_NEW_SMALL_INT(0);
    byte *state;

    self = MP_STATE_PORT(pyb_can_obj_all)[can_id - 1];

    if (fifo_id == CAN_FIFO0) {
        callback = self->rxcallback0;
        state = &self->rx_state0;
    } else {
        callback = self->rxcallback1;
        state = &self->rx_state1;
    }

    switch (*state) {
        case RX_STATE_FIFO_EMPTY:
            __HAL_CAN_DISABLE_IT(&self->can,  (fifo_id == CAN_FIFO0) ? CAN_IT_FMP0 : CAN_IT_FMP1);
            irq_reason = MP_OBJ_NEW_SMALL_INT(0);
            *state = RX_STATE_MESSAGE_PENDING;
            break;
        case RX_STATE_MESSAGE_PENDING:
            __HAL_CAN_DISABLE_IT(&self->can, (fifo_id == CAN_FIFO0) ? CAN_IT_FF0 : CAN_IT_FF1);
            __HAL_CAN_CLEAR_FLAG(&self->can, (fifo_id == CAN_FIFO0) ? CAN_FLAG_FF0 : CAN_FLAG_FF1);
            irq_reason = MP_OBJ_NEW_SMALL_INT(1);
            *state = RX_STATE_FIFO_FULL;
            break;
        case RX_STATE_FIFO_FULL:
            __HAL_CAN_DISABLE_IT(&self->can, (fifo_id == CAN_FIFO0) ? CAN_IT_FOV0 : CAN_IT_FOV1);
            __HAL_CAN_CLEAR_FLAG(&self->can, (fifo_id == CAN_FIFO0) ? CAN_FLAG_FOV0 : CAN_FLAG_FOV1);
            irq_reason = MP_OBJ_NEW_SMALL_INT(2);
            *state = RX_STATE_FIFO_OVERFLOW;
            break;
        case RX_STATE_FIFO_OVERFLOW:
            // This should never happen
            break;
    }

    pyb_can_handle_callback(self, fifo_id, callback, irq_reason);
}
```

```c
typedef enum _rx_state_t {
    RX_STATE_FIFO_EMPTY = 0,
    RX_STATE_MESSAGE_PENDING,
    RX_STATE_FIFO_FULL,
    RX_STATE_FIFO_OVERFLOW,
} rx_state_t;
```

This is how it  should be written

```c
STATIC void can_rx_irq_handler(uint can_id, uint fifo_id) {
    mp_obj_t callback;
    pyb_can_obj_t *self;
    byte *state;

    self = MP_STATE_PORT(pyb_can_obj_all)[can_id - 1];

    if (fifo_id == CAN_FIFO0) {
        callback = self->rxcallback0;
        state = &self->rx_state0;
    } else {
        callback = self->rxcallback1;
        state = &self->rx_state1;
    }

    mp_obj_t irq_reason = MP_OBJ_NEW_SMALL_INT(*state);

    switch (*state) {
        case RX_STATE_FIFO_EMPTY:
            __HAL_CAN_DISABLE_IT(&self->can,  (fifo_id == CAN_FIFO0) ? CAN_IT_FMP0 : CAN_IT_FMP1);
            *state = RX_STATE_MESSAGE_PENDING;
            break;
        case RX_STATE_MESSAGE_PENDING:
            __HAL_CAN_DISABLE_IT(&self->can, (fifo_id == CAN_FIFO0) ? CAN_IT_FF0 : CAN_IT_FF1);
            __HAL_CAN_CLEAR_FLAG(&self->can, (fifo_id == CAN_FIFO0) ? CAN_FLAG_FF0 : CAN_FLAG_FF1);

            *state = RX_STATE_FIFO_FULL;
            break;
        case RX_STATE_FIFO_FULL:
            __HAL_CAN_DISABLE_IT(&self->can, (fifo_id == CAN_FIFO0) ? CAN_IT_FOV0 : CAN_IT_FOV1);
            __HAL_CAN_CLEAR_FLAG(&self->can, (fifo_id == CAN_FIFO0) ? CAN_FLAG_FOV0 : CAN_FLAG_FOV1);
            *state = RX_STATE_FIFO_OVERFLOW;
            break;
        case RX_STATE_FIFO_OVERFLOW:
            // This should never happen
            mp_obj_t irq_reason = MP_OBJ_NEW_SMALL_INT(RX_STATE_FIFO_EMPTY);
            break;
    }
```


I do not believe this line
`            mp_obj_t irq_reason = MP_OBJ_NEW_SMALL_INT(RX_STATE_FIFO_EMPTY);`
should be there but to not break API that is how the function should be written. For readibility purposes this would be better

```c
STATIC void can_rx_irq_handler(uint can_id, uint fifo_id) {
    mp_obj_t callback;
    pyb_can_obj_t *self;
    mp_obj_t irq_reason = MP_OBJ_NEW_SMALL_INT(RX_STATE_FIFO_EMPTY);
    byte *state;

    self = MP_STATE_PORT(pyb_can_obj_all)[can_id - 1];

    if (fifo_id == CAN_FIFO0) {
        callback = self->rxcallback0;
        state = &self->rx_state0;
    } else {
        callback = self->rxcallback1;
        state = &self->rx_state1;
    }

    switch (*state) {
        case RX_STATE_FIFO_EMPTY:
            __HAL_CAN_DISABLE_IT(&self->can,  (fifo_id == CAN_FIFO0) ? CAN_IT_FMP0 : CAN_IT_FMP1);
            irq_reason = MP_OBJ_NEW_SMALL_INT(RX_STATE_FIFO_EMPTY);
            *state = RX_STATE_MESSAGE_PENDING;
            break;
        case RX_STATE_MESSAGE_PENDING:
            __HAL_CAN_DISABLE_IT(&self->can, (fifo_id == CAN_FIFO0) ? CAN_IT_FF0 : CAN_IT_FF1);
            __HAL_CAN_CLEAR_FLAG(&self->can, (fifo_id == CAN_FIFO0) ? CAN_FLAG_FF0 : CAN_FLAG_FF1);
            irq_reason = MP_OBJ_NEW_SMALL_INT(RX_STATE_MESSAGE_PENDING);
            *state = RX_STATE_FIFO_FULL;
            break;
        case RX_STATE_FIFO_FULL:
            __HAL_CAN_DISABLE_IT(&self->can, (fifo_id == CAN_FIFO0) ? CAN_IT_FOV0 : CAN_IT_FOV1);
            __HAL_CAN_CLEAR_FLAG(&self->can, (fifo_id == CAN_FIFO0) ? CAN_FLAG_FOV0 : CAN_FLAG_FOV1);
            irq_reason = MP_OBJ_NEW_SMALL_INT(RX_STATE_FIFO_FULL);
            *state = RX_STATE_FIFO_OVERFLOW;
            break;
        case RX_STATE_FIFO_OVERFLOW:
            // This should never happen
            break;
    }

    pyb_can_handle_callback(self, fifo_id, callback, irq_reason);
}
```

I am not sure what this is 

```c
def cb0(bus, reason):
  if reason == 0:
      print('pending')
  if reason == 1:
      print('full')
  if reason == 2:
      print('overflow')
```

but if it has anything to do with the rx callback it is incorrectly written

If a buffer overrun did occur it would be incorrectly reported as the buffer being empty and the user would not be able to write code to properly address the problem.

Reply:
we are both on the same page as far as how it should work.  the ESP32 while it is a peppy MCU a buffer overrun on the CAN interface can occur and it would be far better to collect the messages inside the interrupt handler if it can be done and pass that message to the scheduler to hand it off to the callback when it can.

That makes more sense to me to do it that way. The memory can be allocated for the number of messages it is able to do this with when the callback gets registered.  an array of message structs can get made and one of the fields in the struct would be for marking the message as read so the interrupt handler would know if they could use that structure for an incoming message.This gives the user some ability to tweak it to suit their needs. I know you cannot allocate memory in an interrupt handler and I feel this would e an idea way to handle it. a static memory allocation for received messages and that allocation is able to be changed by unregistered the callback and registering it again with different number of messages to allocate. 

Or at least give the user the option to do it that way or call the callback directly from the interrupt handler and not schedule the callback this way there is no delay and the user is able to act on it right away.